# **Data validation and cleaning**

## 1. Know Your Data

### *Import required libraries*

In [1]:
import pandas as pd
import numpy as np

### *Loading Dataset*

In [2]:
data = pd.ExcelFile('retail.xlsx')
dfs = []
for sheet in data.sheet_names:
    df = pd.read_excel(data,sheet_name=sheet)
    dfs.append(df)
final_df = pd.concat(dfs,ignore_index=True)

In [3]:
final_df

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                              POSTAGE         1   

                InvoiceDate  Price  Customer ID         Country  
0       2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1       2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2       2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3       2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4       2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                     ...    ...          ...             ...  
1067366 2011-12-09 12:50:00   2.10      12680.0          France  
1067367 2011-12-09 12:50:00   4.15      12680.0          France  
1067368 2011-12-09 12:50:00   4.15      12680.0          France  
1067369 2011-12-09 12:50:00   4.95      12680.0          France  
1067370 2011-12-09 12:50:00  18.00      12680.0          France  

[1067371 rows x 8 columns]

### *Dataset Rows and Column Count*

In [4]:
print(f"Number of Rows:{final_df.shape[0]}")
print(f"Number of Columns:{final_df.shape[1]}")

Number of Rows:1067371
Number of Columns:8


### *DataSet Information*

In [5]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[ns]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 65.1+ MB


### *Duplicate Values*

In [6]:
final_df.duplicated().sum()

np.int64(34335)

### *Missing Values*

In [7]:
# Function for finding perent missing values
def missing_val(data):
    missing_values = data.isna().sum()
    data_len = len(data)
    percent_missing_values = (missing_values/data_len)*100
    return(round(percent_missing_values,2))

In [8]:
final_df.apply(missing_val)

Invoice         0.00
StockCode       0.00
Description     0.41
Quantity        0.00
InvoiceDate     0.00
Price           0.00
Customer ID    22.77
Country         0.00
dtype: float64

## 2. Understanding the Variables

### *Variable Description*

In [9]:
final_df.describe()

Quantity                    InvoiceDate         Price  \
count  1.067371e+06                        1067371  1.067371e+06   
mean   9.938898e+00  2011-01-02 21:13:55.394028544  4.649388e+00   
min   -8.099500e+04            2009-12-01 07:45:00 -5.359436e+04   
25%    1.000000e+00            2010-07-09 09:46:00  1.250000e+00   
50%    3.000000e+00            2010-12-07 15:28:00  2.100000e+00   
75%    1.000000e+01            2011-07-22 10:23:00  4.150000e+00   
max    8.099500e+04            2011-12-09 12:50:00  3.897000e+04   
std    1.727058e+02                            NaN  1.235531e+02   

         Customer ID  
count  824364.000000  
mean    15324.638504  
min     12346.000000  
25%     13975.000000  
50%     15255.000000  
75%     16797.000000  
max     18287.000000  
std      1697.464450

In [10]:
for col in final_df.columns:
    print(f"\n{col}:")
    print(df[col].unique())


Invoice:
[536365 536366 536368 ... 581585 581586 581587]

StockCode:
['85123A' 71053 '84406B' ... '90214U' '47591b' 23843]

Description:
['WHITE HANGING HEART T-LIGHT HOLDER' 'WHITE METAL LANTERN'
 'CREAM CUPID HEARTS COAT HANGER' ... 'lost'
 'CREAM HANGING HEART T-LIGHT HOLDER' 'PAPER CRAFT , LITTLE BIRDIE']

Quantity:
[     6      8      2      3     32      4     24     12     48     18
     20     36     80     64     10    120     96     23      5      1
     -1     50     40    100    192    432    144    288    -12    -24
     16      9    128     25     30     28      7     56     72    200
    600    480     -6     14     -2     11     33     13     -4     -5
     -7     -3     70    252     60    216    384    -10     27     15
     22     19     17     21     34     47    108     52  -9360    -38
     75    270     42    240     90    320   1824    204     69    -36
   -192   -144    160   2880   1400     39    -48    -50     26   1440
     31     82     78     97     98   

### *About dataset*

- The dataset contains 8 Columns namely **Invoice**,**StockCode**,**Description**,**Qunatity**,**InvoiceDate**,**Price**,**CustomerID**,and **Country**.
- This a large data set with **1067371 Rows**.
- There are **34335** duplicate rows in dataset.
- There are **0.41%** missing values in **Description** Column and **22.77%** missing values in **CustomerID** Column
- There are Negative entries in Quantity column and Price column

# 3. Data Wrangling

In [11]:
# Remove Duplicated values
final_df = final_df.drop_duplicates()

In [12]:
# Flag bad debt
final_df['is bad debt'] = (final_df['Description'].str.contains("Adjust bad debt",case = False,na = False) & (final_df['Price']<0))

C:\Users\rg345\AppData\Local\Temp\ipykernel_13876\2431323918.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['is bad debt'] = (final_df['Description'].str.contains("Adjust bad debt",case = False,na = False) & (final_df['Price']<0))


In [13]:
sales_df = final_df[~final_df['is bad debt']].drop('is bad debt',axis=1)

In [14]:
# Checking the variables
sales_df.describe()

Quantity                    InvoiceDate         Price  \
count  1.033031e+06                        1033031  1.033031e+06   
mean   1.007692e+01  2011-01-03 14:30:39.631510784  4.767604e+00   
min   -8.099500e+04            2009-12-01 07:45:00  0.000000e+00   
25%    1.000000e+00            2010-07-05 11:38:00  1.250000e+00   
50%    3.000000e+00            2010-12-09 13:34:00  2.100000e+00   
75%    1.000000e+01            2011-07-27 13:17:00  4.150000e+00   
max    8.099500e+04            2011-12-09 12:50:00  3.897000e+04   
std    1.751981e+02                            NaN  9.283735e+01   

         Customer ID  
count  797885.000000  
mean    15313.062777  
min     12346.000000  
25%     13964.000000  
50%     15228.000000  
75%     16788.000000  
max     18287.000000  
std      1696.466663

In [15]:
# There are negative quantieis with 0 price, indicating inventory adjusment
sales_df[(sales_df['Quantity']<0) & (sales_df['Price']==0)]

Invoice StockCode   Description  Quantity         InvoiceDate  Price  \
263      489464     21733  85123a mixed       -96 2009-12-01 10:52:00    0.0   
283      489463     71477         short      -240 2009-12-01 10:52:00    0.0   
284      489467    85123A   21733 mixed      -192 2009-12-01 10:53:00    0.0   
470      489521     21646           NaN       -50 2009-12-01 11:44:00    0.0   
3114     489655     20683           NaN       -44 2009-12-01 17:26:00    0.0   
...         ...       ...           ...       ...                 ...    ...   
1060794  581210     23395         check       -26 2011-12-07 18:36:00    0.0   
1060796  581212     22578          lost     -1050 2011-12-07 18:38:00    0.0   
1060797  581213     22576         check       -30 2011-12-07 18:38:00    0.0   
1062371  581226     23090       missing      -338 2011-12-08 09:56:00    0.0   
1064386  581422     23169       smashed      -235 2011-12-08 15:24:00    0.0   

         Customer ID         Country  
263              NaN  United Kingdom  
283              NaN  United Kingdom  
284              NaN  United Kingdom  
470              NaN  United Kingdom  
3114             NaN  United Kingdom  
...              ...             ...  
1060794          NaN  United Kingdom  
1060796          NaN  United Kingdom  
1060797          NaN  United Kingdom  
1062371          NaN  United Kingdom  
1064386          NaN  United Kingdom  

[3393 rows x 8 columns]

In [16]:
sales_df['is Inventory Adjustment'] = (
    (sales_df['Quantity'] < 0) & (sales_df['Price'] == 0)
)

In [17]:
sales_df['is Return'] = (
    (sales_df['Quantity']<0) & (sales_df['Price'] > 0)
)

In [18]:
sales_df[(sales_df['is Inventory Adjustment']==False) & (sales_df['is Return']==False)].describe()

Quantity                    InvoiceDate         Price  \
count  1.010535e+06                        1010535  1.010535e+06   
mean   1.133648e+01  2011-01-04 07:06:30.117532928  4.064050e+00   
min    1.000000e+00            2009-12-01 07:45:00  0.000000e+00   
25%    1.000000e+00            2010-07-06 09:51:00  1.250000e+00   
50%    4.000000e+00            2010-12-09 14:27:00  2.100000e+00   
75%    1.200000e+01            2011-07-28 10:33:00  4.130000e+00   
max    8.099500e+04            2011-12-09 12:50:00  2.511109e+04   
std    1.314654e+02                            NaN  5.036675e+01   

         Customer ID  
count  779495.000000  
mean    15320.262918  
min     12346.000000  
25%     13971.000000  
50%     15246.000000  
75%     16794.000000  
max     18287.000000  
std      1695.722988

In [19]:
# Calculating Revenue
sales_df['Revenue'] = sales_df['Quantity'] * sales_df['Price']
sales_df.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  \
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom   
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom   
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom   
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom   
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom   

   is Inventory Adjustment  is Return  Revenue  
0                    False      False     83.4  
1                    False      False     81.0  
2                    False      False     81.0  
3                    False      False    100.8  
4                    False      False     30.0

In [20]:
# Creating Year,weekdays,time,date,month
sales_df['InvoiceYear'] = sales_df['InvoiceDate'].dt.year
sales_df['InvoiceMonth'] = sales_df['InvoiceDate'].dt.month_name()
sales_df['InvoiceDay'] = sales_df['InvoiceDate'].dt.day_name()
sales_df['InvoiceTime'] = sales_df['InvoiceDate'].dt.strftime('%H:%M:%S')
sales_df['InvoiceDate'] = sales_df['InvoiceDate'].dt.strftime('%Y-%m-%d')

In [21]:
# Creating a 'customer status' column for customer identification
conditions = [sales_df['Customer ID'].notna(),(sales_df['Customer ID'].isna())&(~sales_df['is Inventory Adjustment'])]
choices = ['Registered','Not Registered']

sales_df['Customer Status'] = np.select(conditions,choices,default='Unknown')

In [30]:
# Promotional/Free column for items
sales_df['Promotional/Free'] = np.where(
    (sales_df['Quantity']>0) & (sales_df['Price']==0),
    "Promotional",
    "Non-Promotional"
)

In [33]:
sales_df[sales_df['Promotional/Free']=='Promotional']

Invoice StockCode                   Description  Quantity InvoiceDate  \
3161     489659     21350                           NaN       230  2009-12-01   
3731     489781     84292                           NaN        17  2009-12-02   
4674     489825     22076            6 RIBBONS EMPIRE          12  2009-12-02   
5904     489861       DOT                DOTCOM POSTAGE         1  2009-12-02   
6378     489882    35751C                           NaN        12  2009-12-02   
...         ...       ...                           ...       ...         ...   
1060795  581211     22142                         check        14  2011-12-07   
1062442  581234     72817                           NaN        27  2011-12-08   
1063965  581406    46000M  POLYESTER FILLER PAD 45x45cm       240  2011-12-08   
1063966  581406    46000S  POLYESTER FILLER PAD 40x40cm       300  2011-12-08   
1064015  581408     85175                           NaN        20  2011-12-08   

         Price  Customer ID         Country  is Inventory Adjustment  \
3161       0.0          NaN  United Kingdom                    False   
3731       0.0          NaN  United Kingdom                    False   
4674       0.0      16126.0  United Kingdom                    False   
5904       0.0          NaN  United Kingdom                    False   
6378       0.0          NaN  United Kingdom                    False   
...        ...          ...             ...                      ...   
1060795    0.0          NaN  United Kingdom                    False   
1062442    0.0          NaN  United Kingdom                    False   
1063965    0.0          NaN  United Kingdom                    False   
1063966    0.0          NaN  United Kingdom                    False   
1064015    0.0          NaN  United Kingdom                    False   

         is Return  Revenue  InvoiceYear InvoiceMonth InvoiceDay InvoiceTime  \
3161         False      0.0         2009     December    Tuesday    17:39:00   
3731         False      0.0         2009     December  Wednesday    11:45:00   
4674         False      0.0         2009     December  Wednesday    13:34:00   
5904         False      0.0         2009     December  Wednesday    14:50:00   
6378         False      0.0         2009     December  Wednesday    16:22:00   
...            ...      ...          ...          ...        ...         ...   
1060795      False      0.0         2011     December  Wednesday    18:36:00   
1062442      False      0.0         2011     December   Thursday    10:33:00   
1063965      False      0.0         2011     December   Thursday    13:58:00   
1063966      False      0.0         2011     December   Thursday    13:58:00   
1064015      False      0.0         2011     December   Thursday    14:06:00   

        Customer Status Promotional/Free  
3161     Not Registered      Promotional  
3731     Not Registered      Promotional  
4674         Registered      Promotional  
5904     Not Registered      Promotional  
6378     Not Registered      Promotional  
...                 ...              ...  
1060795  Not Registered      Promotional  
1062442  Not Registered      Promotional  
1063965  Not Registered      Promotional  
1063966  Not Registered      Promotional  
1064015  Not Registered      Promotional  

[2621 rows x 17 columns]

In [34]:
# Output File to excel
sales_df.to_excel("Cleaned_data.xlsx",sheet_name="data",index=False)